Adds pull-request standards to the contributing guidelines (#6980)
BeeStation/BeeStation-Hornet: .github/CONTRIBUTING.md

If at any point you are confused or unsure of an asset's license or our policy, ask a Maintainer to help you.

## 7. Pull Request Standards

> You should complete the pull-request template in its entirity.

Reviewer: ```suggestion
You should complete the pull-request template in its entirety.
```

> Any pull-request that does not adequately complete the pull-request template will not be merged and may be closed by maintainers (It is up to the discretion of the maintainers to decide if the PR will be accepted, closed or marked as 'do not merge').

Reviewer: ```suggestion
Any pull-request that does not adequately complete the provided template may be closed or marked 'do not merge' by maintainers.
```

 - Balance changes must indicate that they affect the balance of the game.

>  - Changes must be documented in their entirity, and the full extend of their effects mentioned. (For example, if you change it so all mobs are half speed, don't label the PR as 'monkeys now move twice as slow'). Failing to document the full extent of the changes may result in a repo-ban if the intent of hiding changes is seen as malicious.

Reviewer: ```suggestion
 - Changes must be documented in their entirety including the extent of their effects. (For example, if you change it so all mobs are half speed, don't label the PR as 'monkeys now move twice as slow'). Failing to document the full extent of the changes may result in a repo-ban if the intent of hiding changes is seen as malicious.
```

>  - Changes must be documented in their entirity, and the full extend of their effects mentioned. (For example, if you change it so all mobs are half speed, don't label the PR as 'monkeys now move twice as slow'). Failing to document the full extent of the changes may be seen as malicious and could result in a repo-ban in a worst case scenario.

Reviewer 2: This sounds a bit harsh i am not saying you should not document your changes but saying you can get repo banned just because you didn't properly write down a small change is a bit much also this doesn't really specifiy where you should document those changes after you made the PR(i personally like to put them in commit messages and if its major enough mention them in the about section) neither does it say how well you actually should document them it just provides a bad example no good example.

Reviewer 2: Thinking about it i may have interpreted this a bit to harsh myself but mentioning where to document and what would count as a good example at least sounds like a good idea

Reviewer 3: Something like "repeated violation" would be more appropriate here.

Author: It does say 'could' and that its a worse case scenario rather than you will be banned for this, but I can update it.

>  - Changes must be documented in their entirity, and the full extend of their effects mentioned. (For example, if you change it so all mobs are half speed, don't label the PR as 'monkeys now move twice as slow'). Failing to document the full extent of the changes may be seen as malicious and could result in a repo-ban in a worst case scenario.

Author: ```suggestion
 - Changes must be documented in their entirity, and the full extend of their effects mentioned. (For example, if you change it so all mobs are half speed, don't label the PR as 'monkeys now move twice as slow'). Failing to document the full extent of the changes may result in a repo-ban if the intent of hiding changes is seen as malicious.
```

 - The section labeled 'about this pull request' should state the changes that the pull-request makes.
 - The section labeled "why it's good for the game" should include the reasons behind the changes and how they will be good for the game.

>  - The testing section should contain screenshots, videos or a description showing that the PR works as specified. Pull-requests that ignore this section or are not tested may be closed by maintainers. This applies to small PRs that may seem trivial.

Reviewer: ```suggestion
 - The testing section should contain screenshots, videos, and/or reproducible testing procedures showing that the PR works as specified. Pull-requests that ignore this section, or are not tested, may be closed by maintainers. This applies to small PRs that may seem trivial.
```

>  - The testing section should contain screenshots, videos or a description showing that the PR works as specified. Pull-requests that ignore this section or are not tested may be closed by maintainers. This applies to small PRs that may seem trivial.

Reviewer 2: Now ok we all know this section has been ignored almost entirely since it got introduced(mainly cause its a hassle after hours of work writing down an essay of all possible stuff you did to test it) trough as long as a completly text documentation is (in most cases at least) sufficient i guess its fine.

Reviewer 3: For it to be respected template needs to be improved.
It's currently not really readable and ignored by most.

Author: You just have to put a screenshot in it showing that you have opened it and done at least a basic test on it. An essay isn't required

Reviewer: > (mainly cause its a hassle after hours of work writing down an essay of all possible stuff you did to test it)

A basic set of reproduceable testing procedures is really all you need. Adding a screenshot or a video demonstrating that the thing works is equally acceptable and requires fairly minimal effort.

Reviewer 2: In that case yeah sounds fine

>  - The changelog should include a short summary of the changes made, as well as any other authors that contributed to the pull-request (Include their names after the :cl: tag along with your own)

Reviewer: ```suggestion
 - The changelog should include a short summary of the changes made; including any other authors that contributed to the pull-request (Include their names after the :cl: tag along with your own)
```

If at any point you are confused or unsure of an asset's license or our policy, ask a Maintainer to help you.

## 7. Pull Request Standards

You should complete the pull-request template in its entirety.

Any pull-request that does not adequately complete the provided template may be closed or marked 'do not merge' by maintainers.

>  - Balance changes must indicate that they affect the balance of the game.

Reviewer 4: The wording here is a bit awkward. I'd maybe suggest something like this

```suggestion
 - If you make any changes that affect the balance of the game, they must be mentioned as balance changes. Note that this also applies if you're fixing something, even if the result is what was originally intended.
```

Reviewer: > Note that this also applies if you're fixing something, even if the result is what was originally intended.

This is poorly phrased. It implies that all fixes should be categorized as a balance change. I would drop that entire ending sentence. The alterations to the first sentence are good.

Reviewer 4: It might be poorly phrased, but I feel like we've had cases where significant balance changes were PR'd under the banner of fixes, because the original code's intent seemed to be different than what it actually did. What I mean is, even if the balance changes are the result of a fix, the rule should apply in the same way, treating it as a balance change based on what the difference is.

Reviewer: I think it's fine. Maintainers can edit PR bodies, so in those rare cases we can correct the CL as needed. I don't think it's worth going overly verbose and making it too confusion.

Author: I proposed some other changes, but they went to the bottom of the comments since I didn't do it in this thread.

>  - Balance changes must indicate that they affect the balance of the game.

Author: ```suggestion
 - Any changes that may affect game balance should be documented as a balance change. Bug fixes that affect balance should be marked as a balance change if the bug could reasonably be seen as an intentional feature by somebody that has no knowledge of the code.
```

Author: @KubeRoot @Crossedfall

Reviewer: ```suggestion
 - Any changes that may affect game balance should be documented as a balance change. This also applies to bug fixes which directly alter the game's balance.
```

I think it'd be too difficult to determine what a layman would and wouldn't know is a balance change in this context. Probably better to just be broad.

>  - Changes must be documented in their entirety including the extent of their effects. (For example, if you change it so all mobs are half speed, don't label the PR as 'monkeys now move twice as slow'). Failing to document the full extent of the changes may result in a repo-ban if the intent of hiding changes is seen as malicious.

Reviewer 4: I feel like this is phrased too strongly. There are some things that are difficult to test due to needing complex setup, or infeasible to test fully due to the scope of changes. There are also cases where the person interested in contributing doesn't know how to test things properly (yet).

I think the wording should be eased off a little, and that we should instead require people to specify what they didn't test in those cases.

Reviewer: This bit isn't about testing. It's about documenting changes and being honest about those changes.

Reviewer 4: Whoops, I misclicked. I started writing this earlier, but github dropped my unfinished review, and I didn't double-check when making this comment.

The line this was meant to be commented on is:
```
- The testing section should contain screenshots, videos, and/or reproducible testing procedures showing that the PR works as specified. Pull-requests that ignore this section, or are not tested, may be closed by maintainers. This applies to small PRs that may seem trivial.
```

Reviewer: See: https://github.com/BeeStation/BeeStation-Hornet/pull/6980#discussion_r885039673

>  - The section labeled 'about this pull request' should state the changes that the pull-request makes.

Reviewer 4: I feel like this point is too terse/unverbose, it might as well describe the changelog. I'd suggest either getting rid of it, or something more verbose like

```suggestion
 - The section labeled 'about this pull request' should describe the pull request in detail. This includes the change being made from the player's perspective, any important details about how it was implemented, as well as background on the pull request, such as the source of what might be ported or issues that are being fixed. Note that it should not include images and videos, there is a dedicated section for those.
```

Author: About this pull request should allow for images and videos in it if the images and videos show the changes, I think thats too strong.

Imo the 'about this pull request' section should strictly be the what, the changes how they are in an objective manner, with the reason for implementing them going in the 'Why its good for the game section'.

The main reason for this bit is because all the other sections have a quick description too, about this pull request is pretty self-explanatory.

Reviewer 4: > About this pull request should allow for images and videos in it if the images and videos show the changes, I think thats too strong.

This is a matter of personal opinion, but I was supportive of adding the `Screenshots&Videos` collapsible specifically because those tend to take up a lot of space, making browsing PRs painful. That is, it was my understanding that the collapsible is where screenshots and videos should go in general, and I had the impression we were in agreement about it.

> Imo the 'about this pull request' section should strictly be the what, the changes how they are in an objective manner, with the reason for implementing them going in the 'Why its good for the game section'.

I mostly agree, and I think the suggested changes do not contradict anything you said in this reply. Might be a case of bad wording, but for clarity, what I suggested it should be is:
- The actual changes, based on how they affect the users, in detail.
- Important implementation details. In cases of refactors/new systems/maintenance PRs, this would be the bulk of it, since they won't affect users much. In other PRs, this would only be needed when big/risky changes are being made, or implementation is changed in a way that's important to future coders.
- The background on the PR. It's not meant to be the reason for why it's good for the game, but rather elaboration on how the PR came to be. This is meant to be mainly PRs from other codebases being ported, Issues that are being fixed, previous PRs from this codebase that are being addressed in some ways. That said, if the author wants to cite other background, such as previous discussion about the matter, I feel like this is the good place to do it.

I also want to note that I see the "Why it's good for the game" as being a more strict section - to me, it shouldn't have vague elaboration, but specific reasoning for why it's an improvement. Because of this, anything like points to discuss further that are up to debate would be something to put in "About this pull request".

Author: ```suggestion
 - The section labeled 'about this pull request' should describe the pull request's changes in detail. This includes the changes being made, any important details about how it was implemented, the issues it closes, and links to any other pull requests if code is being ported from another codebase.
```

Reviewer: We're not animals- use the Oxford comma.

Author: what is an oxford comma?

Author: ew that is just wrong

Reviewer: You're objectively wrong. Add the comma, it's far more readable with it.

Author: Fine

Author: That good?

 - The section labeled "why it's good for the game" should include the reasons behind the changes and how they will be good for the game.
 - The testing section should contain screenshots, videos, and/or reproducible testing procedures showing that the PR works as specified. Pull-requests that ignore this section, or are not tested, may be closed by maintainers. This applies to small PRs that may seem trivial.

>  - The changelog should include a short summary of the changes made; including any other authors that contributed to the pull-request (Include their names after the :cl: tag along with your own)

Reviewer 4: I would suggest a different wording here:

```suggestion
 - The changelog should include a short summary of the changes made. If your pull request includes things made by other people, you should list everybody who contributed, including yourself, after the :cl: tag.
```

> If a pull-request requires updates to the wiki, these changes should be made on your user account page (For example: https://wiki.beestation13.com/view/User:Powerfulbacon), so that the original page can be updated on merge.

Author: ```suggestion
If a pull-request requires updates to the wiki, these changes should be made on your user account page (For example: https://wiki.beestation13.com/view/User:PowerfulBacon), so that the original page can be updated on merge.
```

> If a pull-request requires updates to the wiki, these changes should be made on your user account page (For example: https://wiki.beestation13.com/view/User:PowerfulBacon), so that the original page can be updated on merge.

Reviewer 3: "or has needs wiki updates label"

Author: If a pull request has need wiki updates label then it can be assumed the pull-request requires updates to the wiki

Reviewer 3: Nothing wrong with explicitly stating this in contributing

Reviewer: I don't think it's needed.

Author: I also don't think its necessary
